althea_types/src/wifi_info.rs

```diff
@@ -0,0 +1,471 @@
+/// This struct contains information for wifi survey data
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct WifiSurveyData {
+    /// Frequency in MHz
+    pub frequency_mhz: u16,
+    /// Noise in dBm
+    pub noise_dbm: i32,
+    /// Time in ms
+    pub channel_active_time: u64,
+    pub channel_busy_time: u64,
+    pub channel_receive_time: u64,
+    pub channel_transmit_time: u64,
+}
+
+/// This struct contains information for wifi station data
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct WifiStationData {
+    pub station: String,
+    pub inactive_time_ms: u64,
+    pub rx_bytes: u64,
+    pub rx_packets: u64,
+    pub tx_bytes: u64,
+    pub tx_packets: u64,
+    pub tx_retries: u16,
+    pub tx_failed: u16,
+    pub rx_drop_misc: u16,
+    pub signal_dbm: String,
+    pub signal_avg_dbm: String,
+    pub tx_bitrate: f32,
+    pub tx_duration_us: u64,
+    pub rx_bitrate: f32,
+    pub rx_duration_us: u64,
+    pub airtime_weight: u16,
+    pub authorized: bool,
+    pub authenticated: bool,
+    pub associated: bool,
+    pub preamble: String,
+    pub wmm_wme: bool,
+    pub mfp: bool,
+    pub tdls_peer: bool,
+    pub dtim_period: u16,
+    pub beacon_interval: u16,
+    pub short_slot_time: bool,
+    pub connected_time_sec: u16,
+    pub associated_at_boottime_sec: String,
+    pub associated_at_ms: u64,
+    pub current_time_ms: u64,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct WifiDevice {
+    pub name: String,
+    pub survey_data: Vec<WifiSurveyData>,
+    pub station_data: Vec<WifiStationData>,
+}
+
+impl From<Vec<&str>> for WifiSurveyData {
+    fn from(survey: Vec<&str>) -> Self {
+        let mut ret = WifiSurveyData::default();
+        populate_struct_sur(&mut ret, survey);
+        ret
+    }
+}
+
+impl From<Vec<&str>> for WifiStationData {
+    fn from(station: Vec<&str>) -> Self {
+        let mut ret = WifiStationData::default();
+        populate_struct_stat(&mut ret, station);
```

Comment:
can we have this function either be an associated function of WifiStationData so that it's &mut self or just return WifiStationData. Either option seems better than having to take a mutable reference.

In general I'd prefer if the function created the struct from scratch, reduces the chance of someone using an half initialized wifi data at some point in the future.

Reply:
fixed